compiler/fir/checkers/checkers.jvm/src/org/jetbrains/kotlin/fir/analysis/jvm/checkers/expression/FirJavaGenericVarianceViolationTypeChecker.kt

```diff
@@ -0,0 +1,75 @@
+import org.jetbrains.kotlin.types.AbstractTypeChecker
+import org.jetbrains.kotlin.types.model.TypeConstructorMarker
+import org.jetbrains.kotlin.types.model.typeConstructor
+
+object FirJavaGenericVarianceViolationTypeChecker : FirFunctionCallChecker() {
+    private val javaOrigin = setOf(FirDeclarationOrigin.Java, FirDeclarationOrigin.Enhancement)
+    override fun check(expression: FirFunctionCall, context: CheckerContext, reporter: DiagnosticReporter) {
+        val calledFunction = expression.calleeReference.toResolvedCallableSymbol()?.fir as? FirFunction<*> ?: return
+        if (calledFunction.origin !in javaOrigin) {
+            return
+        }
+        val argList = expression.argumentList as? FirArgumentListWithResolutionInfo
+        val paramTypeSubstitutor = argList?.paramTypeSubstitutor ?: return
+        for ((arg, param) in argList.mapping) {
+            val expectedType = paramTypeSubstitutor.substituteOrSelf(param.returnTypeRef.coneType)
+
+            // optimization: if no arguments or flexibility, everything is OK
+            if (expectedType !is ConeFlexibleType || expectedType.typeArguments.isEmpty()) continue
+
+            // Anything is acceptable for raw types
+            if (expectedType is ConeRawType) continue
+
+            val lowerBound = expectedType.lowerBound
+            val upperBound = expectedType.upperBound
+            val typeCtx = context.session.inferenceComponents.ctx
+            val lowerConstructor = lowerBound.typeConstructor(typeCtx)
+            val upperConstructor = upperBound.typeConstructor(typeCtx)
+
+            // Use site variance projection is always the same for flexible types. So there is no need to check if declaration site is the
+            // same.
+            if (lowerConstructor == upperConstructor) return
+
+            val argType = arg.typeRef.coneType
+
+            if (!typeCtx.isTypeConstructorSubClassOf(argType, lowerBound)) return
+
+            if (!AbstractTypeChecker.isSubtypeOf(typeCtx, argType, lowerBound)) {
+                reporter.reportOn(arg.source, FirErrors.JAVA_TYPE_MISMATCH, argType, expectedType, context)
+            }
+        }
+    }
+
+    private fun ConeInferenceContext.isTypeConstructorSubClassOf(subType: ConeKotlinType, superType: ConeKotlinType): Boolean {
+        return isTypeConstructorSubClassOfOrEqual(subType.typeConstructor(), superType.typeConstructor())
+    }
+
+    private fun ConeInferenceContext.isTypeConstructorSubClassOfOrEqual(
+        subTypeConstructor: TypeConstructorMarker,
+        superTypeConstructor: TypeConstructorMarker
+    ): Boolean {
+        if (subTypeConstructor == superTypeConstructor) return true
+        val immediateSuperTypeConstructors = subTypeConstructor.supertypes().map { it.typeConstructor() }
+        if (superTypeConstructor in immediateSuperTypeConstructors) return true
+        return immediateSuperTypeConstructors.any { isTypeConstructorSubClassOfOrEqual(it, superTypeConstructor) }
```

Comment:
Nit: we could avoid materializing (i.e., use of `map`) the list of super type constructor? E.g.,

```kt
for (immediateSuperType in subTypeConstructor.supertypes()) {
  val immediateSuperTypeConstructor = immediateSuperType.typeConstructor()
  if (superTypeConstructor == immediateSuperTypeConstructor) return true
  if (isTypeConstructorSubClassOfOrEqual(immediateSuperTypeConstructor, superTypeConstructor) return true
}
return false
```